### ISSUE TITLE
STI: same property name used in different child classes leads to incorrect generated STI table.

### ISSUE THREAD

**Issue by nicomouss** [mikro-orm/mikro-orm#5813]
### Describe the bug

Let's say my model is made of one base class `Transport` and two other classes `AirTransport` and `RoadTransport` inheriting from that base class:

```ts
abstract class Transport {
    id: number;
    name: string;
}

class AirTransport extends Transport {
    // -----
    // Here a property named "category"
    // -----
    category: AirTransportCategory;
}

class RoadTransport extends Transport {
    // -----
    // Here a property named "category"
    // -----
    category: RoadTransportCategory;
}

enum AirTransportCategory {
    PLANE = "plane",
    HELICOPTER = "helicopter",
}

enum RoadTransportCategory {
    CAR = "car",
    MOTORBIKE = "motorbike",
}
```

Now if I use STI (Single Table Inheritance) using the following `EntitySchema` definitions (I always use `EntitySchema` to keep a clean domain model, but same thing is doable using entity decorators of course):

```ts
const transportSchema = new EntitySchema<Transport>({
    class: Transport,
    abstract: true,
    discriminatorColumn: "type",
    properties: {
        id: {
            fieldName: "id",
            type: Number,
            primary: true,
            autoincrement: true,
        },
        name: {
            fieldName: "name",
            type: String,
        },
    },
    schema: "public",
    tableName: "Transport",
});

const airTransportSchema = new EntitySchema<AirTransport, Transport>({
    class: AirTransport,
    discriminatorValue: "air_transport",
    properties: {
        // -----
        // Here mapping of property "category" of class AirTransport,
        // to column "airTransportCategory" of the STI table.
        // -----
        category: {
            fieldName: "airTransportCategory",
            enum: true,
            items: () => AirTransportCategory,
            nullable: true,
        },
    },
});

const roadTransportSchema = new EntitySchema<RoadTransport, Transport>({
    class: RoadTransport,
    discriminatorValue: "road_transport",
    properties: {
        // -----
        // Here mapping of property "category" of class RoadTransport,
        // to column "roadTransportCategory" of the STI table.
        // -----
        category: {
            fieldName: "roadTransportCategory",
            enum: true,
            items: () => RoadTransportCategory,
            nullable: true,
        },
    },
});
```

Then the table generated by the schema generator is as follow:

```sql
create table "Transport" (
  "id" serial primary key, "name" varchar(255) not null,
  "type" text check ("type" in ('air_transport', 'road_transport')) not null,
  "roadTransportCategory" text check ("roadTransportCategory" in ('car', 'motorbike')) null
);
```

This leads to errors when inserting new entities down the road.

The generated STI table should be something as follow:

```sql
create table "Transport" (
  "id" serial primary key, "name" varchar(255) not null,
  "type" text check ("type" in ('air_transport', 'road_transport')) not null,
  "airTransportCategory" text check ("airTransportCategory" in ('plane', 'helicopter')) null,
  "roadTransportCategory" text check ("roadTransportCategory" in ('car', 'motorbike')) null
);
```

### Reproduction

Please see reproduction code provided above.

The way to get around the issue is to make sure we don't have any property that have the same name in the whole class hierarchy falling into the STI table. But that's kinda a "dirty fix" because in that case, it breaks the OOP/encapsulation principles in the domain entities: meaning when we create a new class, we have to first check if we are not using a property that already exists in another non-related class of the same hierarchy.

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.2.9

### Node.js version

18

### Operating system

Mac os

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by nicomouss** [comment]
I guess this relates to issue #5807.

But strictly speaking, it is not an issue with enums themselves.

It is an issue of having identically named properties (each property having a different type: here different enums, but could be one is "int" and the other is "string") in two different classes of the same STI hierarchy. If that makes sense ...

**Comment by B4nan** [comment]
Closing as duplicate of #5807.

> It is an issue of having identically named properties (each property having a different type: here different enums, but could be one is "int" and the other is "string") in two different classes of the same STI hierarchy. If that makes sense ...

This is a wontfix, it can work only with the same types.

**Comment by B4nan** [comment]
Oh, you have different field names there, I missed that part. I recall we had an issue for that somewhere too, this boils down to how things are internally designed, the root entity has all the properties inlined. I will reopen to check, but this might be actually a wontfix too.

**Comment by nicomouss** [comment]
> Oh, you have different field names there, I missed that part.

@B4nan Yes that's the point I underline with my example. Even with different fieldnames, things do not work out.

**Comment by normalid-awa** [comment]
Will this issue be fix? This actually cause trouble and cause dirty code, as the property name must have redundant prefix to work around this issue.

**Comment by giancarlo-dm** [comment]
I faced this issue multiple times in a project where there were some STIs. I ended up putting a prefix on all props on the subclasses and overriding the the serialized name to not have the prefix in order to keep at least the API clean.

MikroORM could:
- Warn/Error about clashing `fieldNames` in STI entities, prompting us to manually define a `fieldName` on one of them or both.
- Or preferably, to automatically handle clashing field names in STI by automatically adding a prefix based on the `discriminator`.

**Comment by normalid-awa** [comment]
> I faced this issue multiple times in a project where there were some STIs. I ended up putting a prefix on all props on the subclasses and overriding the the serialized name to not have the prefix in order to keep at least the API clean.
> 
> MikroORM could:
> - Warn/Error about clashing `fieldNames` in STI entities, prompting us to manually define a `fieldName` on one of them or both.
> - Or preferably, to automatically handle clashing field names in STI by automatically adding a prefix based on the `discriminator`.

 I don't really think this should throw an error, as it's acceptable and could brings benefits in database design, I think this issue should be treated as a bug and should be fixed. Maybe I could make a PR on this (after reading through the source code).

**Comment by nicomouss** [comment]
@giancarlo-dm Just to clarify things about this issue. The problem I had initially described here is about having **_clashing properties_** (aka properties in typescript/javascript classes), and not **_clashing fieldnames_** (aka column names in db table).

Specifically, it is about **_having identically named properties in two different classes of the same STI hierarchy_**.

For example if you design classes like below, when you have two properties named "category" in two different classes of the class hierarchy, but you still specify different fieldnames for those:

```ts
@entity({abstract: true, discriminatorColumn: "type", ...})
abstract class Transport {
  id: number;
  name: string;
}

@entity({discriminatorValue: "air", ...})
class AirTransport extends Transport {
  @Property({ fieldName: 'air_transport_cat', ... })
  category: AirTransportCategory;
}

@entity({discriminatorValue: "transport", ...})
class RoadTransport extends Transport {
  @Property({ fieldName: 'road_transport_cat', ... })
  category: RoadTransportCategory;
}
```

then Mikroorm does not create two columns **_air_transport_cat_** and **_road_transport_cat_** in the database in the STI table. That's the problem.

When the db schema is created, I'd also prefer MikroOrm to fail with a proper error message. Or just warn the end-user.

**Comment by giancarlo-dm** [comment]
@nicomouss Yes, I know. In my opinion "subclasses" with same properties are not _clashing_, its a design, but on the other hand sharing the same `column` between specialization does not work in some scenarios. Many issues arises beyond the bug we have during hydration.

Your example was only taking into consideration a relationship, which would work just fine because value stored is the PK, which is a number. Although we know that during hydration it currently bugs out and does not hydrate correctly.

Now look at this different scenario (this is what happens to me majority of the time):
- Specialization `A` has a property named `foo` of type `string`
- Specialization `B` has a property also named `foo` but of type `number`

How the migrator would create the db columns if they share the same name but different type? Here we have a clashing `fieldName` with different `type`, that's why MikroORM should at least throw an error because it is syntactically incorrect and will fail at runtime.

Of course, here the developer needs to understand the scenario and should manually define the a custom `fieldName` on each property. But MirkoORM does not warn about it.

Having same properties names is not an issue, rather a design choice by the developer. I agree, MikroORM is bugged in this regard, even if we allow mapping the same `column name` to properties declared in different specializations it does not hydrate things correctly.

**Comment by nicomouss** [comment]
> Your example was only taking into consideration a relationship, which would work just fine because value stored is the PK, which is a number. Although we know that during hydration it currently bugs out and does not hydrate correctly.

There is no relationship nor PKs in my example, just enums.
Which is basically the same thing as having strings or numbers.

> Now look at this different scenario (this is what happens to me majority of the time):
> 
> * Specialization `A` has a property named `foo` of type `string`
> * Specialization `B` has a property also named `foo` but of type `number`

It is no different scenario. I could have written my example like so, to match what you are describing (replace `category` by `foo`):

```ts
@entity(...)
abstract class BaseClass {
  ...
}

@entity(...)
class SpecializationA extends BaseClass {
  @Property({ fieldName: 'categoryA', ... })
  category: string;
}

@entity(...)
class SpecializationB extends BaseClass {
  @Property({ fieldName: 'categoryB', ... })
  category: number;
}
```

So I guess we are basically speaking of the same thing.
I Just wanted to make sure we are on the same page regarding this issue.

> Of course, here the developer needs to understand the scenario and should manually define the a custom fieldName on each property.

That is the key point I am trying to describe in this issue: even if the _developper understands the scenario and manually defines a custom `fieldName` on each property_ (see example above using `categoryA` and `categoryB` as `fieldName` for each property), the generated schema is wrong (only `categoryA` column will be created by MikroOrm in this case).

> Having same properties names is not an issue

That is unfortunately the main issue here. The problem is easily reproductible by doing so:

1. You have two classes `SpecializationA` and `SpecializationB`, inheriting from a base class `BaseClass`. There are properties with the same name in both specializations
2. You want to map those classes using MikroOrm STI functionnality
4. To do so, you specify different `fieldNames` for the identically named properties in order to have different columns generated for each of your properties.
=> an incorrect db schema is generated by MikroOrm (some columns are missing). The reason being that internally when using STI, MikroOrm does not deal well with same properties names.

Thanks for sharing your thoughts on this issue !

**Comment by B4nan** [comment]
The fix is backported to v6 as well, but it only solves the schema generator issue, there are still some problems with the runtime mapping.

**Comment by B4nan** [comment]
Ok, now it should work at runtime too. Backported to v6 as well.
